File: mobile/ios/KinicApp/Utilities/URLNormalizer.swift
// Where: mobile/ios/KinicApp/Utilities/URLNormalizer.swift
// What: URL normalization shared by native capture request generation.
// Why: iOS must match the browser extension and web app URL contract.

import Foundation

enum URLNormalizer {
    static func normalizedHTTPURL(_ url: URL) throws -> URL {
        guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              components.scheme == "http" || components.scheme == "https",
              components.host != nil else {
            throw URLNormalizerError.unsupportedURL
        }
        components.fragment = nil
        guard let normalized = components.url else {
            throw URLNormalizerError.unsupportedURL
        }
        return normalized
    }
}

File: mobile/android/app/src/main/java/xyz/kinic/android/URLNormalizer.kt
// Where: mobile/android/app/src/main/java/xyz/kinic/android/URLNormalizer.kt
// What: URL normalization shared by manual and Android share captures.
// Why: Android request generation must match iOS by accepting only http(s) and removing fragments.

package xyz.kinic.android

import java.net.URI

object URLNormalizer {
    fun normalizedHttpUrl(raw: String): URI =
        normalizedHttpUrl(URI(raw.trim()))

    fun normalizedHttpUrl(uri: URI): URI {
        val scheme = uri.scheme?.lowercase()
        val host = uri.host
        if ((scheme != "http" && scheme != "https") || host.isNullOrBlank()) {
            throw URLNormalizerException.UnsupportedUrl
        }
        return URI(
            scheme,
            uri.userInfo,
            host,
            uri.port,
            uri.path,
            uri.query,
            null,
        )
    }
}

sealed class URLNormalizerException(message: String) : IllegalArgumentException(message) {
    data object UnsupportedUrl : URLNormalizerException("unsupported URL")
}
